rclrs/src/wait/guard_condition.rs

```diff
@@ -0,0 +1,186 @@
+///     &context,
+///     move || {
+///         atomic_bool_for_closure.store(true, Ordering::Relaxed);
+///     },
+/// ));
+///
+/// let mut ws = WaitSet::new(0, 1, 0, 0, 0, 0, &context)?;
+/// ws.add_guard_condition(Arc::clone(&gc))?;
+///
+/// // Trigger the guard condition, firing the callback and waking the wait set being waited on, if any.
+/// gc.trigger()?;
+///
+/// // The provided callback has now been called.
+/// assert_eq!(atomic_bool.load(Ordering::Relaxed), true);
+///
+/// // The wait call will now immediately return.
+/// ws.wait(Some(std::time::Duration::from_millis(10)))?;
+///
+/// # Ok::<(), RclrsError>(())
+/// ```
+pub struct GuardCondition {
+    /// The rcl_guard_condition_t that this struct encapsulates.
+    pub(crate) rcl_guard_condition: Mutex<rcl_guard_condition_t>,
+    /// An optional callback to call when this guard condition is triggered.
+    callback: Option<Box<dyn Fn() + Send + Sync>>,
+    /// A flag to indicate if this guard condition has already been assigned to a wait set.
+    pub(crate) in_use_by_wait_set: Arc<AtomicBool>,
+}
+
+impl Drop for GuardCondition {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: No precondition for this function (besides passing in a valid guard condition)
+            rcl_guard_condition_fini(&mut *self.rcl_guard_condition.lock().unwrap());
+        }
+    }
+}
+
+impl PartialEq for GuardCondition {
+    fn eq(&self, other: &Self) -> bool {
+        // Because GuardCondition controls the creation of the rcl_guard_condition, each unique GuardCondition should have a unique
+        // rcl_guard_condition. Thus comparing equality of this member should be enough.
+        std::ptr::eq(
+            &self.rcl_guard_condition.lock().unwrap().impl_,
+            &other.rcl_guard_condition.lock().unwrap().impl_,
+        )
+    }
+}
+
+impl Eq for GuardCondition {}
+
+// SAFETY: rcl_guard_condition is the only member that doesn't implement Send, and it is designed to be accessed from other threads
+unsafe impl Send for rcl_guard_condition_t {}
+
+impl GuardCondition {
+    /// Creates a new guard condition with no callback.
+    pub fn new(context: &Context) -> Self {
+        Self::new_with_rcl_context(&mut context.rcl_context_mtx.lock().unwrap(), None::<fn()>)
+    }
+
+    /// Creates a new guard condition with a callback.
+    pub fn new_with_callback<F>(context: &Context, callback: F) -> Self
+    where
+        F: Fn() + Send + Sync + 'static,
+    {
+        Self::new_with_rcl_context(&mut context.rcl_context_mtx.lock().unwrap(), Some(callback))
+    }
+
+    /// Creates a new guard condition by providing the rcl_context_t and an optional callback.
+    /// Note this function enables calling `Node::create_guard_condition`[1] without providing the Context separately
+    ///
+    /// [1]: Node::create_guard_condition
+    pub(crate) fn new_with_rcl_context<F>(context: &mut rcl_context_t, callback: Option<F>) -> Self
```

Comment:
Nitpick: If this took a `Box<dyn Fn>`, we could make this non-generic, which is a little better for compile times and you wouldn't need the `::<fn()>` (though it's a nice trick!). But it's not really important, take it or leave it.

Reply:
I was thinking the same thing actually. I'll make this change real quick!